### PR TITLE
WIP: Multiple sol

### PR DESCRIPTION
**Pull request by EtashGuha**

### CODE REVIEW

**Comment by mattmilten** [comment]
@CGraczyk, could you please help out here?

**Comment by EtashGuha** [comment]
Feel free to message me if anything is needed.

**Comment by mattmilten** [comment]
Hi @EtashGuha, could you please update your branch (merge in the master branch or do a rebase) so the pull request doesn't have conflicts anymore?
Thanks!

**Comment by EtashGuha** [comment]
@mattmilten Done! it should not have any merge conflicts!

File: src/pyscipopt/scip.pxd [hunk 1]
```diff
@@ -1608,6 +1633,11 @@ cdef extern from "scip/cons_countsols.h":
     SCIP_RETCODE SCIPcount(SCIP* scip)
     SCIP_RETCODE SCIPsetParamsCountsols(SCIP* scip)
     SCIP_Longint SCIPgetNCountedSols(SCIP* scip, SCIP_Bool* valid)
+    SCIP_RETCODE SCIPincludeConshdlrCountsols   (   SCIP *  scip    )   
+    void SCIPgetCountedSparseSols(SCIP* scip, SCIP_VAR*** vars, int* nvars, SCIP_SPARSESOL*** sols, int* nsols);
+    SCIP_RETCODE SCIPcount  (   SCIP *  scip    );
```
Comment (CGraczyk) [inline src/pyscipopt/scip.pxd:1638]: Please check for duplicates before commiting.

**Review by CGraczyk — COMMENTED** [review]
Please delete empty lines and use the existing style format. It seems the conflict resolution produced duplicate code.

File: src/pyscipopt/scip.pxd [hunk 2]
```diff
@@ -656,6 +673,13 @@ cdef extern from "scip/scip.h":
                                      SCIP_BOUNDTYPE* boundtypes,
                                      int* nbranchvars,
                                      int branchvarssize)
+    void SCIPnodeGetAncestorBranchings(SCIP_NODE* node,
+                                     SCIP_VAR** branchvars,
+                                     SCIP_Real* branchbounds,
+                                     SCIP_BOUNDTYPE* boundtypes,
+                                     int* nbranchvars,
+                                     int branchvarssize)
+
```
[the remark this answers is not in the document]
Comment (EtashGuha) [reply src/pyscipopt/scip.pxd]: Done

File: src/pyscipopt/scip.pxd [hunk 3]
```diff
@@ -1653,13 +1683,30 @@ cdef extern from "scip/pub_lp.h":
     SCIP_Real SCIPcolGetPrimsol(SCIP_COL* col)
     SCIP_Real SCIPcolGetLb(SCIP_COL* col)
     SCIP_Real SCIPcolGetUb(SCIP_COL* col)
+
```
[the remark this answers is not in the document]
Comment (EtashGuha) [reply src/pyscipopt/scip.pxd]: Done

File: src/pyscipopt/scip.pxd [hunk 4]
```diff
@@ -1671,10 +1718,298 @@ cdef extern from "scip/scip_tree.h":
...
+cdef extern from "scip/branch_vanillafullstrong.h":
+    SCIP_RETCODE SCIPincludeBranchruleVanillafullstrong(SCIP* scip)
+    SCIP_RETCODE SCIPgetVanillafullstrongData(SCIP* scip,
+                                              SCIP_VAR*** cands,
+                                              SCIP_Real** candscores,
+                                              int* ncands,
+                                              int* npriocands,
+                                              int* bestcand)
+
+
```
[the remark this answers is not in the document]
Comment (EtashGuha) [reply src/pyscipopt/scip.pxd]: It is unnecessary. Done.

File: src/pyscipopt/scip.pyx [hunk 5]
```diff
@@ -4112,6 +4116,56 @@ cdef class Model:
...
         return SCIPgetNSols(self._scip)
 
+    def getSparseSols(self):
+        """
+
+        :return Solutions: solutions in the original variable space
+        
+        based off of https://www.scipopt.org/doc/html/cons__countsols_8c_source.php#l02693
+
+        """
```
[the remark this answers is not in the document]
Comment (EtashGuha) [reply src/pyscipopt/scip.pyx]: Done.

File: src/pyscipopt/scip.pyx [hunk 6]
```diff
@@ -4358,6 +4414,13 @@ cdef class Model:
 
     # Parameter Methods
 
+    def count(self):
+        print("In correct count")
+        PY_SCIP_CALL(SCIPsetParamsCountsols(self._scip))
+        PY_SCIP_CALL(SCIPsetBoolParam(self._scip, "constraints/countsols/collect", 1));
+        PY_SCIP_CALL(SCIPsetLongintParam(self._scip, "constraints/countsols/sollimit",5) );
+        PY_SCIP_CALL(SCIPcount(self._scip));
+
```
[the remark this answers is not in the document]
Comment (EtashGuha) [reply src/pyscipopt/scip.pyx]: Fixed in new commit. Done.

File: src/pyscipopt/scip.pxd (hunk 1 above)
Comment (EtashGuha) [reply src/pyscipopt/scip.pxd]: Done.

File: src/pyscipopt/scip.pxd [hunk 7]
```diff
@@ -627,6 +642,8 @@ cdef extern from "scip/scip.h":
     SCIP_RETCODE SCIPsetObjIntegral(SCIP* scip)
     SCIP_Real SCIPgetLocalOrigEstimate(SCIP* scip)
     SCIP_Real SCIPgetLocalTransEstimate(SCIP* scip)
+    SCIP_RETCODE SCIPgetProbvarLinearSum(SCIP*                 scip,SCIP_VAR**            vars,SCIP_Real*            scalars,int*                  nvars, int                   varssize,SCIP_Real*            constant,int*                  requiredsize,SCIP_Bool             mergemultiples)
+    SCIP_Bool SCIPsparseSolGetNextSol(SCIP_SPARSESOL*       sparsesol,SCIP_Longint*         sol, int                   nvars)
```
[the remark this answers is not in the document]
Comment (EtashGuha) [reply src/pyscipopt/scip.pxd]: Done.

File: src/pyscipopt/scip.pxd [hunk 8]
```diff
@@ -403,7 +418,7 @@ cdef extern from "scip/scip.h":
     ctypedef struct SCIP_DIVESET:
         pass
 
-    ctypedef struct SCIP_HASHMAP:
+    ctypedef struct SCIP_HASHMAP:
```
[the remark this answers is not in the document]
Comment (EtashGuha) [reply src/pyscipopt/scip.pxd]: Done.

File: src/pyscipopt/scip.pxd [hunk 9]
```diff
@@ -656,6 +673,12 @@ cdef extern from "scip/scip.h":
                                      SCIP_BOUNDTYPE* boundtypes,
                                      int* nbranchvars,
                                      int branchvarssize)
+    void SCIPnodeGetAncestorBranchings(SCIP_NODE* node,
+                                     SCIP_VAR** branchvars,
+                                     SCIP_Real* branchbounds,
+                                     SCIP_BOUNDTYPE* boundtypes,
+                                     int* nbranchvars,
+                                     int branchvarssize)
```
Comment (CGraczyk) [inline src/pyscipopt/scip.pxd:681]: How is this connected to the functionality of getSparseSols() ?

File: src/pyscipopt/scip.pxd [hunk 10]
```diff
@@ -281,8 +281,23 @@ cdef extern from "scip/scip.h":
...
+    ctypedef struct SCIP_SET:
         pass
 
+    ctypedef struct SCIP_PROB:
+        pass
+
+    ctypedef struct SCIP_LP:
+        SCIP_Real objsqrnorm
+
+
```
Comment (CGraczyk) [inline src/pyscipopt/scip.pxd:300]: @mattmilten is this style conform?

File: src/pyscipopt/scip.pxd [hunk 11]
```diff
@@ -794,6 +817,7 @@ cdef extern from "scip/scip.h":
     SCIP_SOL* SCIPgetBestSol(SCIP* scip)
     SCIP_Real SCIPgetSolVal(SCIP* scip, SCIP_SOL* sol, SCIP_VAR* var)
     SCIP_RETCODE SCIPwriteVarName(SCIP* scip, FILE* outfile, SCIP_VAR* var, SCIP_Bool vartype)
+    SCIP_RETCODE SCIPgetTransformedVar(SCIP* scip, SCIP_VAR* var, SCIP_VAR** transvar)
```
Comment (CGraczyk) [inline src/pyscipopt/scip.pxd:820]: Could you explain where this is used for your intended functionality?

File: src/pyscipopt/scip.pxd [hunk 12]
```diff
@@ -1660,6 +1686,21 @@ cdef extern from "scip/pub_lp.h":
...
+    int SCIProwGetAge(SCIP_ROW* row)
+    SCIP_Real SCIProwGetLPActivity(SCIP_ROW* row,
+                                   SCIP_SET* set,
+                                   SCIP_STAT* stat,
+                                   SCIP_LP* lp)
+    SCIP_Real SCIProwGetObjParallelism(SCIP_ROW* row,
+                                       SCIP_SET* set,
+                                       SCIP_LP* lp)
+    SCIP_Real SCIPcolGetObj(SCIP_COL *col)
+
```
Comment (CGraczyk) [inline src/pyscipopt/scip.pxd:1703]: Could you explain where this is used for your intended functionality?

File: src/pyscipopt/scip.pxd [hunk 13]
```diff
@@ -1671,6 +1712,12 @@ cdef extern from "scip/scip_tree.h":
     SCIP_NODE* SCIPgetBestNode(SCIP* scip)
     SCIP_NODE* SCIPgetBestboundNode(SCIP* scip)
     SCIP_RETCODE SCIPrepropagateNode(SCIP* scip, SCIP_NODE* node)
+    int SCIPgetFocusDepth(SCIP *scip)
+    int SCIPgetPlungeDepth(SCIP *scip)
+    int SCIPgetEffectiveRootDepth(SCIP* scip)
+    int SCIPgetNNodesLeft(SCIP* scip)
+    int SCIPgetCutoffdepth(SCIP* scip)
+    int SCIPgetRepropdepth(SCIP* scip)
```
Comment (CGraczyk) [inline src/pyscipopt/scip.pxd:1720]: Could you explain where this is used for your intended functionality?

File: src/pyscipopt/scip.pyx [hunk 14]
```diff
@@ -4120,9 +4174,11 @@ cdef class Model:
         nsols = SCIPgetNSols(self._scip)
         sols = []
 
+        
         for i in range(nsols):
             sols.append(Solution.create(self._scip, _sols[i]))
 
+
```
Comment (CGraczyk) [inline src/pyscipopt/scip.pyx:4181]: Please try to avoid adding empty lines.

**Review by CGraczyk — COMMENTED** [review]
Could you please explain the usage of the different functions you added in your code? some of it seems disconnected from the intended functionality. 
Maybe you could add what the diffs in your commit are intended to do, so we can work together and add things step by step.

File: src/pyscipopt/scip.pyx [hunk 15]
```diff
@@ -4112,6 +4115,57 @@ cdef class Model:
...
 
+    def getSparseSols(self):
+        """gets all counted solutions in the original variable space by taking sparse solutions in transformed variable space and 
+        converting them to solutions to solutions in original variable space
+
+        :return Solution solutions: solutions in the original variable space
+        
+        based off of https://www.scipopt.org/doc/html/cons__countsols_8c_source.php#l02693
+
+        """
```
Comment (CGraczyk) [inline src/pyscipopt/scip.pyx:4126]:
```suggestion
        """
        gets all counted solutions in the original variable space by taking sparse solutions in the
        transformed variable space and converting them to solutions in the original variable space

        :return Solution 
        solutions: solutions in the original variable space
        """
```

Is this correct?

File: src/pyscipopt/scip.pxd (hunk 13 above)
Comment (EtashGuha) [reply src/pyscipopt/scip.pxd]: It has none. Sorry. I cleaned it up more. It should be good now :)

File: src/pyscipopt/scip.pxd (hunk 12 above)
Comment (EtashGuha) [reply src/pyscipopt/scip.pxd]: It has none. I cleaned it up. :)

File: src/pyscipopt/scip.pxd (hunk 11 above)
Comment (EtashGuha) [reply src/pyscipopt/scip.pxd]: I have not used it yet. Sorry. :)

File: src/pyscipopt/scip.pxd (hunk 10 above)
Comment (EtashGuha) [reply src/pyscipopt/scip.pxd]: I cleaned it up. :)

File: src/pyscipopt/scip.pxd (hunk 9 above)
Comment (EtashGuha) [reply src/pyscipopt/scip.pxd]: It doesnt. I cleaned it up :)

File: src/pyscipopt/scip.pxd [hunk 16]
```diff
@@ -1608,6 +1610,7 @@ cdef extern from "scip/cons_countsols.h":
     SCIP_RETCODE SCIPcount(SCIP* scip)
     SCIP_RETCODE SCIPsetParamsCountsols(SCIP* scip)
     SCIP_Longint SCIPgetNCountedSols(SCIP* scip, SCIP_Bool* valid)
+    void SCIPgetCountedSparseSols(SCIP* scip, SCIP_VAR*** vars, int* nvars, SCIP_SPARSESOL*** sols, int* nsols);
 
 cdef extern from "scip/paramset.h":
```
Comment (EtashGuha) [inline src/pyscipopt/scip.pxd:1616]: This function fetches the counted sparse solutions

File: src/pyscipopt/scip.pxd [hunk 17]
```diff
@@ -627,6 +627,8 @@ cdef extern from "scip/scip.h":
     SCIP_RETCODE SCIPsetObjIntegral(SCIP* scip)
     SCIP_Real SCIPgetLocalOrigEstimate(SCIP* scip)
     SCIP_Real SCIPgetLocalTransEstimate(SCIP* scip)
+    SCIP_RETCODE SCIPgetProbvarLinearSum(SCIP* scip, SCIP_VAR** vars, SCIP_Real* scalars, int* nvars, int varssize, SCIP_Real* constant, int* requiredsize, SCIP_Bool mergemultiples)
```
Comment (EtashGuha) [inline src/pyscipopt/scip.pxd:630]: This transforms variable in original variable space into transformed space.

File: src/pyscipopt/scip.pxd [hunk 18]
```diff
@@ -627,6 +627,8 @@ cdef extern from "scip/scip.h":
...
+    SCIP_Bool SCIPsparseSolGetNextSol(SCIP_SPARSESOL* sparsesol, SCIP_Longint* sol, int nvars)
```
Comment (EtashGuha) [inline src/pyscipopt/scip.pxd:631]: This function fills in the next counted solution into the sol parameter

**Comment by CGraczyk** [comment]
Your code is now merged into the feature branch - we can develop the actual functionality including a test to make sure the function is working correctly in a new pull request.

**Pull request merged**